hivemind/p2p/p2p_daemon_bindings/control.py

```diff
@@ -93,6 +103,100 @@ async def listen(self) -> AsyncIterator["ControlClient"]:
+                    self.pending_calls[call_id].set_result(resp.callUnaryResponse.result)
+                elif call_id in self.pending_calls and resp.callUnaryResponse.HasField("error"):
+                    remote_exc = RemoteException(str(resp.callUnaryResponse.error))
+                    self.pending_calls[call_id].set_exception(remote_exc)
+                else:
+                    logger.debug(f"received unexpected unary call")
+
+            elif resp.HasField("requestHandling"):
+                asyncio.create_task(self._handle_persistent_request(resp.requestHandling))
+                pass
+
+    async def _write_to_persistent_conn(self, writer: asyncio.StreamWriter):
+        while True:
+            msg = await self.pending_messages.get()
+            await write_pbmsg(writer, msg)
+
+    async def _handle_persistent_request(self, request):
+        assert request.proto in self.unary_handlers
+
+        try:
+            remote_id = PeerID(request.peer)
+            response_payload: bytes = await self.unary_handlers[request.proto](request.data, remote_id)
+            response = p2pd_pb.CallUnaryResponse(callId=request.callId, result=response_payload)
+        except Exception as e:
+            response = p2pd_pb.CallUnaryResponse(callId=request.callId, error=repr(e))
+
+        await self.pending_messages.put(
+            p2pd_pb.Request(type=p2pd_pb.Request.SEND_RESPONSE_TO_REMOTE, sendResponseToRemote=response)
+        )
+
+    async def _handler(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
+        pb_stream_info = p2pd_pb.StreamInfo()  # type: ignore
+        await read_pbmsg_safe(reader, pb_stream_info)
+        stream_info = StreamInfo.from_protobuf(pb_stream_info)
+        try:
+            handler = self.handlers[stream_info.proto]
+        except KeyError as e:
+            # should never enter here... daemon should reject the stream for us.
+            writer.close()
+            raise DispatchFailure(e)
+        await handler(stream_info, reader, writer)
+
+    async def _ensure_persistent_conn(self):
+        if not self._pers_conn_open:
+            reader, writer = await self.daemon_connector.open_persistent_connection()
+            asyncio.create_task(self._read_from_persistent_conn(reader))
+            asyncio.create_task(self._write_to_persistent_conn(writer))
+            self._pers_conn_open = True  # TODO FIXME
+
+    async def add_unary_handler(self, proto: str, handler: TUnaryHandler):
+        await self._ensure_persistent_conn()
+
+        add_unary_handler_req = p2pd_pb.AddUnaryHandlerRequest(proto=proto)
+        req = p2pd_pb.Request(
+            type=p2pd_pb.Request.ADD_UNARY_HANDLER,
+            addUnaryHandler=add_unary_handler_req,
+        )
+        await self.pending_messages.put(req)
+
+        if self.unary_handlers.get(proto):
+            raise ValueError(f"Handler for protocol {proto} already assigned")
+        self.unary_handlers[proto] = handler
```

Comment:
Let's move these lines before `await self.pending_messages.put(req)` because:

  1. We want to raise ValueError before registering the duplicate handler in the daemon.
  2. In case if some badly written coroutine will take asyncio loop for a long time, other peers may start sending requests to the handler before `await self.pending_messages.put(req)` returns control to this function.